packages/stream-tests/src/__tests__/cross-node-indexing.test.ts

```diff
@@ -0,0 +1,172 @@
+
+  beforeAll(async () => {
+    process.env.CERAMIC_ENABLE_EXPERIMENTAL_INDEXING = 'true'
+
+    ipfs1 = await createIPFS()
+    ipfs2 = await createIPFS()
+
+    // Temporarily start a Ceramic node and use it to create the Model that will be used in the
+    // rest of the tests.
+    ceramic1 = await createCeramic(ipfs1)
+
+    model = await Model.create(ceramic1, MODEL_DEFINITION)
+    midMetadata = { model: model.id }
+
+    await ceramic1.close()
+  })
+
+  afterAll(async () => {
+    await ipfs1.stop()
+    await ipfs2.stop()
+  })
+
+  beforeEach(async () => {
+    const indexingDirectory1 = await tmp.tmpName()
+    await fs.mkdir(indexingDirectory1)
+    ceramic1 = await createCeramic(ipfs1, {
+      indexing: {
+        db: `sqlite://${indexingDirectory1}/ceramic.sqlite`,
+        models: [model.id.toString()],
+        allowQueriesBeforeHistoricalSync: true,
+      },
+    })
+    const indexingDirectory2 = await tmp.tmpName()
+    await fs.mkdir(indexingDirectory2)
+    ceramic2 = await createCeramic(ipfs2, {
+      indexing: {
+        db: `sqlite://${indexingDirectory2}/ceramic.sqlite`,
+        models: [model.id.toString()],
+        allowQueriesBeforeHistoricalSync: true,
+      },
+    })
+  }, 30 * 1000)
+
+  afterEach(async () => {
+    await ceramic1.close()
+    await ceramic2.close()
+  })
+
+  test('Index is synced across nodes', async () => {
+    const doc1 = await ModelInstanceDocument.create(ceramic1, CONTENT0, midMetadata, {
+      anchor: false,
+    })
+
+    // TODO: Once we support subscriptions, use a subscription to wait for the stream to show up
+    // in the index, instead of this race-prone sleep.
+    await TestUtils.delay(5 * 1000)
+
+    let resultObj = await ceramic2.index.queryIndex({ model: model.id, first: 100 })
+    let results = extractDocuments(ceramic2, resultObj)
+
+    expect(results.length).toEqual(1)
+    expect(results[0].id.toString()).toEqual(doc1.id.toString())
+    expect(results[0].content).toEqual(CONTENT0)
+    expect(results[0].state).toEqual(doc1.state)
+
+    // Now update an existing stream and create a new stream and make sure the index updates properly
+    await doc1.replace(CONTENT1, { anchor: false })
+    const doc2 = await ModelInstanceDocument.create(ceramic1, CONTENT2, midMetadata, {
+      anchor: false,
+    })
+
+    // TODO: Once we support subscriptions, use a subscription to wait for the stream to show up
+    // in the index, instead of this race-prone sleep.
+    await TestUtils.delay(5 * 1000)
+
+    resultObj = await ceramic2.index.queryIndex({ model: model.id, first: 100 })
+    results = extractDocuments(ceramic2, resultObj)
+
+    expect(results.length).toEqual(2)
+    expect(results[0].id.toString()).toEqual(doc1.id.toString())
+    expect(results[0].content).toEqual(CONTENT1)
+    // TODO: Why do I need to serialize state for this to work?
+    expect(StreamUtils.serializeState(results[0].state)).toEqual(
```

Comment:
I don't understand why this is necessary in this one case but not the other times I compare states in this test.  I think it has something to do with how a StreamID inside the state object is Memoized.  @ukstv I know you were looking into some weirdness around memoization before.